src/main/kotlin/io/libp2p/core/security/noise/NoiseXXSecureChannel.kt

```diff
@@ -0,0 +1,207 @@
+            super.channelActive(ctx)
+            channelRegistered(ctx)
+            logger.debug("Activation complete")
+        }
+
+        private var activated = false
+        private var flagRemoteVerified = false
+        private var flagRemoteVerifiedPassed = false
+        private var aliceSplit: CipherState? = null
+        private var bobSplit: CipherState? = null
+        private var cipherStatePair: CipherStatePair? = null
+
+        override fun channelRead(ctx: ChannelHandlerContext, msg1: Any) {
+            logger.debug("Starting channelRead0")
+            val msg = if (msg1 is ByteArray) {
+                msg1
+            } else {
+                (msg1 as ByteBuf).array()
+            }
+
+            channelActive(ctx)
+
+            if (role == HandshakeState.RESPONDER && flagRemoteVerified && !flagRemoteVerifiedPassed) {
+                logger.error("Responder verification of Remote peer id has failed")
+                throw Exception("Responder verification of Remote peer id has failed")
+            }
+
+            // if we are here, we are still in handshake setup phase
+
+            // we always read from the wire when it's the next action to take
+            val payload = ByteArray(65535)
+            var payloadLength = 0
+            if (handshakestate.action == HandshakeState.READ_MESSAGE) {
+                payloadLength = handshakestate.readMessage(msg, 0, msg.size, payload, 0)
+            }
+
+            if (role == HandshakeState.RESPONDER && !flagRemoteVerified) {
+                // the self-signed remote pubkey and signature would be retrieved from the first Noise payload
+                val inp = Spipe.Exchange.parseFrom(payload.copyOfRange(0, payloadLength))
+                // validate the signature
+                val inpub = unmarshalPublicKey(inp.epubkey.toByteArray())
+                val verification = inpub.verify(inp.epubkey.toByteArray(), inp.signature.toByteArray())
+
+                flagRemoteVerified = true
+                if (verification) {
+                    logger.debug("Remote verification passed")
+                    flagRemoteVerifiedPassed = true
+                } else {
+                    logger.error("Remote verification failed")
+                    flagRemoteVerifiedPassed = false // being explicit about it
+                    throw Exception("Responder verification of Remote peer id has failed")
+                    // throwing exception for early exit of protocol and for application to handle
+                }
+            }
+
+            // after reading messages and setting up state, write next message onto the wire
+            if (handshakestate.action == HandshakeState.WRITE_MESSAGE) {
+                val sndmessage = ByteArray(65535)
+                val sndmessageLength: Int
+                sndmessageLength = handshakestate.writeMessage(sndmessage, 0, null, 0, 0)
+                ctx.writeAndFlush(sndmessage.copyOfRange(0, sndmessageLength))
```

Comment:
You need to convert message to `ByteBuf` here (`ByteArray.toByteBuf`)

Reply:
done